File: backend/services/resumeParser.js
const fs = require('fs');

/**
 * Common technical skills dictionary for auto-tagging.
 */
const SKILL_KEYWORDS = [
  'JavaScript', 'TypeScript', 'React', 'React.js', 'Node.js', 'Node', 'Python',
  'Java', 'C++', 'C#', 'Go', 'Golang', 'Rust', 'Ruby', 'PHP', 'Swift', 'Kotlin',
  'HTML', 'CSS', 'Tailwind', 'Sass', 'Bootstrap',
  'SQL', 'MySQL', 'PostgreSQL', 'MongoDB', 'Redis', 'SQLite',
  'Express', 'Next.js', 'Vite', 'Django', 'Flask', 'Spring',
  'AWS', 'GCP', 'Azure', 'Docker', 'Kubernetes', 'Terraform',
  'Git', 'GitHub', 'CI/CD', 'Linux', 'REST', 'RESTful', 'GraphQL',
  'Playwright', 'Puppeteer', 'Selenium', 'Jest', 'Cypress',
  'LLM', 'AI', 'Machine Learning', 'Deep Learning', 'PyTorch', 'TensorFlow'
];

/**
 * Parse text content from a PDF file and extract structured candidate fields.
 *
 * @param {string} filePath - Path to the PDF resume
 * @param {Object} currentCandidate - Existing candidate profile to overlay on
 * @returns {Promise<Object>} Updated candidate profile object
 */
async function parseAndExtractResume(filePath, currentCandidate = {}) {
  try {
    const dataBuffer = fs.readFileSync(filePath);
    const uint8Array = new Uint8Array(dataBuffer);

    let text = '';
    try {
      const pdfModule = require('pdf-parse');
      if (pdfModule.PDFParse) {
        const parser = new pdfModule.PDFParse(uint8Array);
        const data = await parser.getText();
        if (typeof data === 'string') text = data;
        else if (data && data.text) text = data.text;
        else if (data && Array.isArray(data.pages)) {
          text = data.pages.map((p) => (p.text ? p.text : JSON.stringify(p))).join('\n');
        }
      } else if (typeof pdfModule === 'function') {
        const data = await pdfModule(dataBuffer);
        text = data.text || '';
      }
    } catch (e) {
      console.warn('[ResumeParser] Primary PDFParse attempt failed, trying raw text scan:', e.message);
      text = dataBuffer.toString('utf8');
    }

    if (!text.trim()) {
      return currentCandidate;
    }

    const lines = text
      .split('\n')
      .map((l) => l.trim())
      .filter((l) => l.length > 0);

    const extracted = { ...currentCandidate };

    // 1. Extract Email
    const emailMatch = text.match(/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/i);
    if (emailMatch) {
      extracted.email = emailMatch[0];
    }

    // 2. Extract Phone Number
    const phoneMatch = text.match(/(?:\+\d{1,3}[\s-]?)?\(?\d{3}\)?[\s-]?\d{3}[\s-]?\d{4}|\+?\d{10,13}/);
    if (phoneMatch) {
      extracted.phone = phoneMatch[0];
    }

    // 3. Extract LinkedIn URL
    const linkedinMatch = text.match(/(?:https?:\/\/)?(?:www\.)?linkedin\.com\/in\/[a-zA-Z0-9_-]+\/?/i);
    if (linkedinMatch) {
      extracted.linkedin = linkedinMatch[0].startsWith('http') ? linkedinMatch[0] : `https://${linkedinMatch[0]}`;
    }

    // 4. Extract GitHub URL
    const githubMatch = text.match(/(?:https?:\/\/)?(?:www\.)?github\.com\/[a-zA-Z0-9_-]+\/?/i);
    if (githubMatch) {
      extracted.github = githubMatch[0].startsWith('http') ? githubMatch[0] : `https://${githubMatch[0]}`;
    }

    // 5. Extract Candidate Name (First non-empty line usually states candidate name)
    const possibleNameLine = lines.find(
      (l) =>
        l.length < 50 &&
        !l.includes('@') &&
        !l.includes('http') &&
        !/\d{5,}/.test(l) &&
        !l.toLowerCase().includes('resume') &&
        !l.toLowerCase().includes('curriculum')
    );

    if (possibleNameLine) {
      // Split header line (e.g. "TOUSIF RAZA | Bengaluru" or "TOUSIF RAZA")
      const cleanLine = possibleNameLine.split('|')[0].trim();
      const parts = cleanLine.replace(/[^a-zA-Z\s]/g, '').trim().split(/\s+/);
      if (parts.length >= 2) {
        extracted.firstName = capitalize(parts[0]);
        extracted.lastName = parts.slice(1).map(capitalize).join(' ');
      } else if (parts.length === 1 && parts[0].length > 1) {
        extracted.firstName = capitalize(parts[0]);
      }
    }

    // 6. Extract Location (City/Country pattern or lines with city names)
    const locationMatch = text.match(/([A-Z][a-z]+(?:\s[A-Z][a-z]+)?),\s*([A-Z][a-z]+|[A-Z]{2})/);
    if (locationMatch) {
      extracted.location = locationMatch[0];
    } else {
      const cityMatch = lines.find((l) => /bengaluru|bangalore|mumbai|delhi|pune|hyderabad|chennai|san francisco|new york|london|paris|remote/i.test(l));
      if (cityMatch) {
        extracted.location = cityMatch.split('|')[0].trim();
      }
    }

    // 7. Extract Skills matching dictionary
    const foundSkills = new Set(extracted.skills || []);
    for (const skill of SKILL_KEYWORDS) {
      const escaped = skill.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      const regex = new RegExp(`\\b${escaped}\\b`, 'i');
      if (regex.test(text)) {
        foundSkills.add(skill);
      }
    }
    if (foundSkills.size > 0) {
      extracted.skills = Array.from(foundSkills);
    }

    // 8. Extract Education snippet if present
    const eduLine = lines.find((l) =>
      /b\.e\.|b\.tech|b\.s\.|m\.s\.|master|bachelor|degree|university|college/i.test(l)
    );
    if (eduLine) {
      extracted.education = eduLine.slice(0, 100);
    }

    // 9. Extract Current Title / Role
    const titleLine = lines.find((l) =>
      /engineer|developer|intern|manager|analyst|consultant|architect|specialist/i.test(l)
    );
    if (titleLine) {
      extracted.currentTitle = titleLine.split('|')[0].trim().slice(0, 60);
    }

    console.log(`[ResumeParser] Parsed profile for: ${extracted.firstName} ${extracted.lastName}`);
    return extracted;
  } catch (err) {
    console.error('[ResumeParser] PDF parsing warning:', err.message);
    return currentCandidate;
  }
}

function capitalize(str = '') {
  if (!str) return '';
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

module.exports = { parseAndExtractResume };

File: python_scraper/resume_parser.py
import os
import re
import json
import sys

def parse_pdf(file_path):
    text = ""
    try:
        import pdfplumber
        with pdfplumber.open(file_path) as pdf:
            for page in pdf.pages:
                t = page.extract_text()
                if t:
                    text += t + "\n"
    except Exception as e:
        sys.stderr.write(f"[PythonResumeParser] pdfplumber error: {e}, falling back to pypdf\n")

    if not text.strip():
        try:
            import pypdf
            reader = pypdf.PdfReader(file_path)
            for page in reader.pages:
                t = page.extract_text()
                if t:
                    text += t + "\n"
        except Exception as e:
            sys.stderr.write(f"[PythonResumeParser] pypdf error: {e}\n")

    return text

def parse_docx(file_path):
    text = ""
    try:
        import docx
        doc = docx.Document(file_path)
        for p in doc.paragraphs:
            if p.text:
                text += p.text + "\n"
        for table in doc.tables:
            for row in table.rows:
                row_text = [cell.text.strip() for cell in row.cells if cell.text.strip()]
                if row_text:
                    text += " | ".join(row_text) + "\n"
    except Exception as e:
        sys.stderr.write(f"[PythonResumeParser] docx parse error: {e}\n")
    return text

def parse_txt(file_path):
    try:
        with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
            return f.read()
    except Exception as e:
        sys.stderr.write(f"[PythonResumeParser] txt parse error: {e}\n")
        return ""

SKILL_DICTIONARY = [
    "JavaScript", "TypeScript", "React", "React.js", "Node.js", "Node", "Python",
    "Java", "C++", "C#", "Go", "Golang", "Rust", "Ruby", "PHP", "Swift", "Kotlin",
    "HTML", "CSS", "Tailwind", "TailwindCSS", "Sass", "Bootstrap",
    "SQL", "MySQL", "PostgreSQL", "MongoDB", "Redis", "SQLite", "DynamoDB", "Snowflake", "BigQuery",
    "Express", "Next.js", "Vite", "Django", "Flask", "Spring", "FastAPI", "Nuxt.js",
    "AWS", "GCP", "Azure", "Docker", "Kubernetes", "Terraform", "Ansible", "CI/CD",
    "Git", "GitHub", "Linux", "REST", "RESTful", "GraphQL", "gRPC", "WebSockets",
    "Playwright", "Puppeteer", "Selenium", "Jest", "Cypress", "PyTest",
    "LLM", "AI", "Machine Learning", "Deep Learning", "PyTorch", "TensorFlow", "OpenAI",
    "LangChain", "LlamaIndex", "RAG", "Vector DB", "Pinecone", "Milvus", "Chroma",
    "Pandas", "NumPy", "Scikit-Learn", "SciPy", "Keras", "Hugging Face", "Transformers",
    "Kafka", "Spark", "Airflow", "Microservices", "System Design", "OOP"
]

def extract_candidate_data(file_path, current_candidate=None):
    if current_candidate is None:
        current_candidate = {}

    if not os.path.exists(file_path):
        return current_candidate

    ext = os.path.splitext(file_path)[1].lower()

    if ext == ".pdf":
        text = parse_pdf(file_path)
    elif ext in [".docx", ".doc"]:
        text = parse_docx(file_path)
    else:
        text = parse_txt(file_path)

    if not text.strip():
        return current_candidate

    extracted = dict(current_candidate)
    lines = [l.strip() for l in text.split("\n") if l.strip()]

    # 1. Email
    email_match = re.search(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}', text)
    if email_match:
        extracted["email"] = email_match.group(0)

    # 2. Phone
    phone_match = re.search(r'(?:\+\d{1,3}[\s-]?)?\(?\d{3}\)?[\s-]?\d{3}[\s-]?\d{4}|\+?\d{10,13}', text)
    if phone_match:
        extracted["phone"] = phone_match.group(0)

    # 3. LinkedIn & GitHub
    li_match = re.search(r'(?:https?:\/\/)?(?:www\.)?linkedin\.com\/in\/[a-zA-Z0-9_-]+\/?', text, re.IGNORECASE)
    if li_match:
        url = li_match.group(0)
        extracted["linkedin"] = url if url.startswith("http") else f"https://{url}"

    gh_match = re.search(r'(?:https?:\/\/)?(?:www\.)?github\.com\/[a-zA-Z0-9_-]+\/?', text, re.IGNORECASE)
    if gh_match:
        url = gh_match.group(0)
        extracted["github"] = url if url.startswith("http") else f"https://{url}"

    # 4. Name extraction
    possible_name = next((l for l in lines[:10] if len(l) < 50 and "@" not in l and "http" not in l and not re.search(r'\d{5,}', l) and "resume" not in l.lower()), None)
    if possible_name:
        clean_name = possible_name.split("|")[0].strip()
        parts = [p for p in re.sub(r'[^a-zA-Z\s]', '', clean_name).strip().split() if p]
        if len(parts) >= 2:
            extracted["firstName"] = parts[0].capitalize()
            extracted["lastName"] = " ".join([p.capitalize() for p in parts[1:]])
        elif len(parts) == 1:
            extracted["firstName"] = parts[0].capitalize()

    # 5. Location extraction
    loc_match = re.search(r'([A-Z][a-z]+(?:\s[A-Z][a-z]+)?),\s*([A-Z][a-z]+|[A-Z]{2})', text)
    if loc_match:
        extracted["location"] = loc_match.group(0)
    else:
        city_line = next((l for l in lines[:15] if re.search(r'bengaluru|bangalore|mumbai|delhi|pune|hyderabad|chennai|san francisco|new york|london|paris|remote', l, re.IGNORECASE)), None)
        if city_line:
            extracted["location"] = city_line.split("|")[0].strip()

    # 6. Skills extraction
    found_skills = set(extracted.get("skills", []))
    for skill in SKILL_DICTIONARY:
        escaped = re.escape(skill)
        if re.search(rf'\b{escaped}\b', text, re.IGNORECASE):
            found_skills.add(skill)
    if found_skills:
        extracted["skills"] = list(found_skills)

    # 7. Current Title / Role
    title_line = next((l for l in lines if re.search(r'engineer|developer|intern|manager|analyst|consultant|architect|specialist', l, re.IGNORECASE)), None)
    if title_line:
        extracted["currentTitle"] = title_line.split("|")[0].strip()[:60]

    # 8. Education Snippet
    edu_line = next((l for l in lines if re.search(r'b\.e\.|b\.tech|b\.s\.|m\.s\.|master|bachelor|degree|university|college', l, re.IGNORECASE)), None)
    if edu_line:
        extracted["education"] = edu_line[:100]

    return extracted

if __name__ == "__main__":
    if len(sys.argv) > 1:
        target_path = sys.argv[1]
        base_candidate = {}
        if len(sys.argv) > 2:
            try:
                base_candidate = json.loads(sys.argv[2])
            except Exception:
                pass
        result = extract_candidate_data(target_path, base_candidate)
        print(json.dumps(result, indent=2))
